Libraries/MLXVLM/Models/Paligemma.swift

```diff
@@ -0,0 +1,743 @@
+        self._languageModel.wrappedValue = Language.LanguageModel(config.textConfiguration)
+        self._multiModalProjector.wrappedValue = PaliGemmaMultiModalProjector(
+            config.visionConfiguration)
+    }
+
+    private func inputEmbeddings(inputIds: MLXArray, pixelValues: MLXArray?, mask: MLXArray) -> (
+        MLXArray, MLXArray
+    ) {
+        guard let pixelValues else {
+            return (inputIds, mask)
+        }
+
+        let inputEmbedding = languageModel.model.embedTokens(inputIds)
+        let (hiddenState, _, _) = self.visionModel(
+            pixelValues.transposed(0, 2, 3, 1).asType(inputEmbedding.dtype),
+            outputHiddenStates: true
+        )
+
+        var imageFeatures = hiddenState[.newAxis, .ellipsis].asType(pixelValues.dtype)
+        imageFeatures = multiModalProjector(imageFeatures)
+
+        return prepareInputsForMultimodal(
+            imageFeatures: imageFeatures, inputEmbedding: inputEmbedding,
+            inputIds: inputIds, attentionMask: mask)
+    }
+
+    private func prepareInputsForMultimodal(
+        imageFeatures: MLXArray, inputEmbedding: MLXArray, inputIds: MLXArray,
+        attentionMask: MLXArray
+    ) -> (MLXArray, MLXArray) {
+        let embedDimension = imageFeatures.dim(2)
+        let (batchSize, sequenceLength) = inputIds.shape2
+        var scaledImageFeatures = imageFeatures / pow(Float(config.hiddenSize), 0.5)
+        var finalEmbedding = zeros([batchSize, sequenceLength, embedDimension])
+
+        let textMask = (inputIds .!= config.imageTokenIndex) & (inputIds .!= config.padTokenId)
+        let imageMask = inputIds .== config.imageTokenIndex
+        let padMask = inputIds .== config.padTokenId
+
+        // expand masks to match embedding dimension
+        var textMaskExpanded = expandedDimensions(textMask, axis: -1)
+        textMaskExpanded = repeated(textMaskExpanded, count: embedDimension, axis: -1)
+        var padMaskExpanded = expandedDimensions(padMask, axis: -1)
+        padMaskExpanded = repeated(padMaskExpanded, count: embedDimension, axis: -1)
+
+        // insert padding and text token embeddings
+        finalEmbedding = which(textMaskExpanded, inputEmbedding, finalEmbedding)
+        finalEmbedding = which(padMaskExpanded, zeros(like: finalEmbedding), finalEmbedding)
+
+        let padSize = finalEmbedding.dim(1) - scaledImageFeatures.dim(1)
+        scaledImageFeatures = padded(scaledImageFeatures, widths: [0, .init((0, padSize)), 0])
+
+        // insert image embeddings - the image mask is always less or equal to the sentence in length
+        var imageMaskExpanded = expandedDimensions(imageMask, axis: -1)
+        imageMaskExpanded = repeated(imageMaskExpanded, count: embedDimension, axis: -1)
+        finalEmbedding = which(imageMaskExpanded, scaledImageFeatures, finalEmbedding)
+
+        finalEmbedding = which(padMaskExpanded, zeros(like: finalEmbedding), finalEmbedding)
```

Comment:
In python it's better to do:

`mx.where(mask, array, 0.0)` since the `0` will be broadcast and inherit the type of `array`. I think the same is true in Swift?